sphinx/source/performance.rst

```diff
@@ -0,0 +1,103 @@
+                verify_384_1k |       1 |     2.966 | 2966300 | 52.408 |      337.1
+         verify_256k1_mbed_1k |       1 |     1.722 | 1721700 | 30.419 |      580.8
+       verify_256k1_bitc_1k * |       1 |     0.060 |   59600 |  1.053 |    16778.5
+              verify_384_100k |       1 |     3.162 | 3161701 | 55.860 |      316.3
+       verify_256k1_mbed_100k |       1 |     2.111 | 2110901 | 37.295 |      473.7
+     verify_256k1_bitc_100k * |       1 |     0.398 |  398000 |  7.032 |     2512.6
+    ===============================================================================
+    hash:
+    ===============================================================================
+       Name (baseline is *)   |   Dim   |  Total ms |  ns/op  |Baseline| Ops/second
+    ===============================================================================
+               hash_384_1byte |       1 |     0.001 |     600 |  0.750 |  1666666.7
+        hash_256k1_mbed_1byte |       1 |     0.001 |     600 |  0.750 |  1666666.7
+      hash_256k1_bitc_1byte * |       1 |     0.001 |     800 |      - |  1250000.0
+                  hash_384_1k |       1 |     0.003 |    3100 |  3.875 |   322580.6
+           hash_256k1_mbed_1k |       1 |     0.004 |    4200 |  5.250 |   238095.2
+         hash_256k1_bitc_1k * |       1 |     0.004 |    4500 |  5.625 |   222222.2
+                hash_384_100k |       1 |     0.223 |  223000 |278.750 |     4484.3
+         hash_256k1_mbed_100k |       1 |     0.340 |  340100 |425.125 |     2940.3
+       hash_256k1_bitc_100k * |       1 |     0.340 |  340400 |425.500 |     2937.7
+    ===============================================================================
+
+This compares the signing, verification, and hashing performance of different cryptographic curve implementations in CCF. The `bitcoin-core implementation <bitcoin_256k1>`_ of elliptic curve secp256k1 is significantly faster than other supported curves, particularly when verification becomes a bottleneck.
+
+End-to-end performance tests
+----------------------------
+
+The end-to-end service performance tests can also be from the CCF build directory:
+
+.. code-block:: bash
+
+    ./tests.sh -VV -L "perf"
+
+Each of these tests creates a temporary CCF service on the local machine, then sends a high volume of transactions to measure peak and average throughput. The python test wrappers will print summary statistics including a transaction rate histogram when the test completes. These statistics can be retrieved from any CCF service via the ``getMetrics`` RPC.
+
+For a finer grained view of performance the clients in these tests can also dump the precise times each transaction was sent and its response received, for later analysis. The ``samples`` folder contains a ``plot_tx_times`` Python script demonstrating how this data may be plotted. The following plot was produced by this script, showing 1,000 transactions sent to the `SmallBank`_ sample application:
```

Comment:
It is not clear how `plot_tx_times` should be run. Does `./tests.sh -VV -L "perf"` have to be run beforehand? I think we should add an example on how to run this, from a fresh build.

Reply:
I've added an example invocation of `plot_tx_times`. I'm wary of outlining precise repro steps - this (whole document, but this section in particular) is a gesture in the direction of an investigative tool, rather than a step-by-step guide. If we want something general and directly useable for app developers, I think it needs to be a dashboard/tooling around the `getMetrics` RPC.